File: src/action_space/cozmo_actions.py
#!/usr/bin/env python3

"""
This file implements a generic action space for cozmo to use given certain contraints
"""
import cozmo
from cozmo.util import distance_mm

import math
import numpy as np

class Action(object):
    """
    An action class that stores the following information:
        linear velocity
        angular velocity
        duration
    """
    def __init__(self, lin_vel, ang_vel, duration):
        """
        Parameters
        ----------
        lin_vel : float
            linear velocity, in millimeters/s
        ang_vel : float
            angular velocity, in millimeters/s
        duration : float
            duration, in seconds
        """

        self.lin_vel = lin_vel
        self.ang_vel = ang_vel
        self.duration = duration

    def __str__(self):
        return "Linear Velocity: %s, Angular Velocity: %s, Duration: %s" % \
                (self.lin_vel, self.ang_vel, self.duration)

    def __repr__(self):
        return "Action(%s, %s, %s)" % (self.lin_vel, self.ang_vel, self.duration)

class ActionSpace(object):
    """
    An action space class that generates possible actions 
    for cozmo to execute given the following constraints:
        minimum, maximum, and number of samples for:
            linear velocity, in millimeters/s
            angular velocity, in millimeters/s
            duration, in seconds
    
    Note: Linear Velocity > 0 represents forward movement, 
                          < 0 represents backward movement,
                          == 0 represents no movement in this direction
          Angular Velocity > 0 represents clockwise rotation, 
                           < 0 represents counterclockwise, 
                           == 0 represents no movement in this direction
          
          Linear Velocity is defined as the velocity of cozmo's left and right wheel 
            moving in the same direction
          Angular Velocity is defined as the velocity of cozmo's left and right wheel
            moving in opposite directions, for example, for an angular velocity of 100 mm/s,
            this will correspond to the left wheel moving forward at 100 mm/s and the right 
            wheel moving backwards at 100 mm/s

    Methods
    -------
    apply_action(index)
        applies an action from the generated action space
    view_action_space()
        output all the actions in the action space with the corresponding indices
    """
    def __init__(self, cozmo, lin_min, lin_max, lin_samples, ang_min, ang_max, ang_samples, dur_min, dur_max, dur_samples):
        """
        Parameters
        ----------
        cozmo : cozmo.robot
            the cozmo SDK robot handle
        lin_min : float
            the minimum linear velocity to be generated, in millimeters/s
        lin_max : float
            the maximum linear velocity to be generated, in millimeters/s
        lin_samples : int
            the number of samples to be generated for linear velocity
        ang_min : float
            the minimum angular velocity to be generated, in millimeters/s
        ang_max : float
            the maximum angular velocity to be generated, in millimeters/s
        ang_samples : int
            the number of samples to be generated for angular velocity
        dur_min : float
            the minimum duration to be generated, in seconds
        dur_max : float
            the maximum duration to be generated, in seconds
        dur_samples : int
            the number of samples to be generated for angular velocity
        """

        self.cozmo = cozmo
        self.lin_min = lin_min
        self.lin_max = lin_max
        self.lin_samples = lin_samples
        self.ang_min = ang_min
        self.ang_max = ang_max
        self.ang_samples = ang_samples
        self.dur_min = dur_min
        self.dur_max = dur_max
        self.dur_samples = dur_samples
        self._generate_actions()
        
    def apply_action(self, action_id):
        """
        Applies an action from the action space based on the id

        Parameters
        ----------
        action_id : int
            the id of the action in the action space

        Raises
        ------
        IndexError
            if the id is not a valid id in the action space
        """

        if action_id >= len(self.actions) or action_id < 0:
            raise IndexError(action_id, ' is an invalid id\n Enter an id between 0 and ', len(self.actions) - 1)

        action = self.actions[action_id]
        print('Applying action: ', action)
        left_wheel = action.lin_vel + action.ang_vel
        right_wheel = action.lin_vel - action.ang_vel
        self.cozmo.drive_wheels(left_wheel, right_wheel, duration=action.duration)
    
    def view_action_space(self):
        """
        Outputs all the actions from the action space with their corresponding indices
        """

        index = 0
        for action in self.actions:
            if index % 5 == 0:
                print()
            print(index, ' : ', action)
            index += 1

    def _generate_actions(self):
        """
        Helper function to generate the action space
        """

        self.actions = []
        lin_choices = self._create_choices(self.lin_min, self.lin_max, self.lin_samples, True)
        ang_choices = self._create_choices(self.ang_min, self.ang_max, self.ang_samples, True)
        dur_choices = self._create_choices(self.dur_min, self.dur_max, self.dur_samples, False)

        for dur_choice in dur_choices:
            for lin_choice in lin_choices:
                for ang_choice in ang_choices:
                    if not lin_choice == ang_choice == 0 or dur_choice == 0:
                        action = Action(lin_choice, ang_choice, dur_choice)
                        self.actions.append(action)

    def _create_choices(self, start, stop, num, include_zero):
        """
        Helper function to generate choices, generates [num] number of choices
        from [start] to [stop]

        Parameters
        ----------
        start : float
           the starting value of the sequence
        stop : float
            the end value of the sequence
        num : int
            number of samples to generate
        include_zero : bool
            True to add zero to choices, False to not
            This allows for 0 velocity in either the linear or angular direction

        Returns a list of choices
        """

        choices = np.linspace(start, stop, num)
        if include_zero:
            choices = np.unique(np.insert(choices, 0, 0, axis=0))
        return [self._truncate(val, 3) for val in list(choices)]


    def _truncate(self, number, digits) -> float:
        """
        Helper function to truncate floats to specified number of decimal places

        Parameters
        ----------
        number : float
            the number to truncate
        digits : int
            the number of decimal places to keep

        Returns the truncated number
        """
        stepper = 10.0 ** digits
        return math.trunc(stepper * number) / stepper

def cozmo_run(robot: cozmo.robot):
    action = ActionSpace(robot, 10, 100, 5, 10, 100, 5, 1, 5, 5)
    action.view_action_space()
    action.apply_action(174)

if __name__ == '__main__':
    cozmo.run_program(cozmo_run)

File: src/action_space/cozmo_actions.cpp
#include <algorithm>
#include <cfloat>
#include <iostream>
#include <iterator>
#include <list>
#include <math.h>
#include <numeric>
#include <vector>
using namespace std;

struct Action
{
    double lin_vel;
    double ang_vel;
    double duration;
};

struct Pose
{
    double x;
    double y;
    double z;
    double angle_z;
};

struct Point
{
    double x;
    double y;
};

struct Object_Oriented_Action
{
    Pose pose;
    Action action;
};

// Linear interpolation following MATLAB linspace
vector<double> generate_samples(double min, double max, size_t N) {
    double step = (max - min) / static_cast<double>(N - 1);
    vector<double> samples(N);
    vector<double>::iterator i;
    double val;
    for (i = samples.begin(), val = min; i != samples.end(); ++i, val += step) {
        *i = val;
    }
    return samples;
}

// Utility function to generate [num] number of choices from [start] to [stop]
// include_zero : True to add zero to choices, False to not
//                This allows for 0 verlocity in either the linear or angular direction
vector<double> create_choices(double start, double stop, int num, bool include_zero)
{
    double step = (stop - start) / (num - 1);
    vector<double> choices = generate_samples(start, stop, num);
    
    if (include_zero) {
        bool contains_zero = false;
        if (find(choices.begin(), choices.end(), 0) != choices.end()) {
            contains_zero = true;
        }
        if (!contains_zero) {
            choices.insert(choices.begin(), 0);
        }
    }
    return choices;
}

class GenericActionSpace
{
    public:
        GenericActionSpace(double lin_min,
                           double lin_max,
                           double lin_samples,
                           double ang_min,
                           double ang_max,
                           double ang_samples,
                           double dur_min,
                           double dur_max,
                           double dur_samples)
        : lin_min(lin_min),
          lin_max(lin_max),
          lin_samples(lin_samples),
          ang_min(ang_min),
          ang_max(ang_max),
          ang_samples(ang_samples),
          dur_min(dur_min),
          dur_max(dur_max),
          dur_samples(dur_samples)
        { generate_actions(); }

        list<Action> get_action_space() {
            return actions;
        }

    private:
        double lin_min{ 0 };
        double lin_max{ 0 };
        double lin_samples{ 0 };
        double ang_min{ 0 };
        double ang_max{ 0 };
        double ang_samples{ 0 };
        double dur_min{ 0 };
        double dur_max{ 0 };
        double dur_samples{ 0 };
        list<Action> actions;

        void generate_actions() {
            vector<double> lin_choices = create_choices(lin_min, lin_max, lin_samples, true);
            vector<double> ang_choices = create_choices(ang_min, ang_max, ang_samples, true);
            vector<double> dur_choices = create_choices(dur_min, dur_max, dur_samples, false);

            for (const auto& dur_choice : dur_choices) {
                for (const auto& lin_choice : lin_choices) {
                    for (const auto& ang_choice : ang_choices) {
                        if (!(lin_choice == ang_choice && lin_choice == 0) || dur_choice == 0) {
                            Action a {lin_choice, ang_choice, dur_choice};
                            actions.push_back(a);
                        }
                    }
                }
            }
        }
};

class ObjectOrientedActionSpace
{
    public:
        ObjectOrientedActionSpace(Pose pose,
                                  int samples,
                                  double lin_min,
                                  double lin_max,
                                  double lin_samples,
                                  double dur_min,
                                  double dur_max,
                                  double dur_samples)
        : pose(pose),
          samples(samples),
          lin_min(lin_min),
          lin_max(lin_max),
          lin_samples(lin_samples),
          dur_min(dur_min),
          dur_max(dur_max),
          dur_samples(dur_samples)
        { generate_actions();
        }
        
        list<Object_Oriented_Action> get_action_space() {
            return actions;
        }

    private:
        Pose pose{ 0, 0, 0, 0 };
        double samples{ 0 };
        double lin_min{ 0 };
        double lin_max{ 0 };
        double lin_samples{ 0 };
        double dur_min{ 0 };
        double dur_max{ 0 };
        double dur_samples{ 0 };
        list<Object_Oriented_Action> actions;

        Point cube_offset(double offset, double angle) {
            return Point{offset * cos(angle), offset * sin(angle)};
        }
        
        /*
        Helper function to find the location of all 4 sides of the cube

        Parameters
        ----------
        angle : the angle of the cube, in radians
        
        returns a sorted list of the angle of each of the 4 sides where index
            0 corresponds to front of cube
            1 corresponds to left of cube
            2 corresponds to back of cube
            3 corresponds to right of cube
        */
        vector<double> find_sides(double angle)
        {
            vector<double> sides;
            vector<double> ordered_sides;
            
            sides.push_back(angle);
            for (size_t i = 0; i < 3; i++) {
                // Adding in clockwise order
                angle -= M_PI / 2;
                if (angle < -M_PI) {
                    angle = 2 * M_PI + angle;
                }
                sides.push_back(angle);
            }
            
            int front_idx = nearest_zero(sides);
            ordered_sides.push_back(sides[front_idx]);
            int idx = (front_idx + 1) % 4;
            while (idx != front_idx) {
                ordered_sides.push_back(sides[idx]);
                idx = (idx + 1) % 4;
            }
            return ordered_sides;
        }
        

        void generate_actions(double h_offset=40, double v_offset=60)
        {
            vector<Pose> locations = generate_offsets(h_offset, v_offset);
            GenericActionSpace gen_space (lin_min, lin_max, lin_samples, 0, 0, 0, dur_min, dur_max, dur_samples);
            list<Action> gen_actions = gen_space.get_action_space();
            for (const auto& location : locations) {
                for (const auto& action : gen_actions) {
                    Object_Oriented_Action ooa{location, action};
                    actions.push_back(ooa);
                }
            }
        }

        /*
        Helper function to generate cube offset positions

        Parameters
        ----------
        h_offset : the max horizontal offset from the center of the edge of the cube, in millimeters
        v_offset : the vertical offset away from the center of the cube, in millimeters
        */
        vector<Pose> generate_offsets(double h_offset, double v_offset)
        {
            vector<double> choices;
            if (samples == 1) {
                choices.push_back(0);
            } else {
                choices = create_choices(-h_offset, h_offset, samples, true);
            }
            vector<double> cube_sides = find_sides(pose.angle_z);
            vector<Pose> locations;
            for (size_t i = 0; i < cube_sides.size(); i++) {
                double side = cube_sides[i];
                for (const auto& choice : choices) {
                    Point offset = cube_offset(v_offset, side);
                    Pose location;
                    if (i == 0) { // front of cube
                        location = Pose{pose.x - offset.x, pose.y - offset.y - choice, pose.z, side};
                    } else if (i == 1) { // left of cube
                        location = Pose{pose.x - offset.x - choice, pose.y - offset.y, pose.z, side};
                    } else if (i == 2) { // back of cube
                        location = Pose{pose.x - offset.x, pose.y + offset.y + choice, pose.z, side};
                    } else { // right of cube 
                        location = Pose{pose.x + offset.x + choice, pose.y - offset.y, pose.z, side};
                    }
                    locations.push_back(location);
                }
            }
            return locations;
        }

        /** Helper function to find the value closest to zero in a list,
            used in find_sides to identify which angle of the cube
            is the front

            Note: in case the corner of the cube is perfectly in align with cozmo
            and there is no closest side, we choose the right side to be the front

            returns the index of the value closest to zero
        */
        int nearest_zero(vector<double> values) {
            int nearest = 0;
            for (size_t i = 1; i < values.size(); i++) {
                if (abs(values[i]) < abs(values[nearest])) {
                    nearest = i;
                }
                if (values[i] == M_PI / 2) {
                    return i;
                }
            }
            return nearest;
        }
};


int main() {
    GenericActionSpace gen_space (10, 100, 5, 10, 100, 5, 1, 5, 5);
    list<Action> gen_actions = gen_space.get_action_space();

    // for (const auto& action : actions) {
    //     cout << "Linear Velocity: " << action.lin_vel << " Angular Velocity: " << action.ang_vel << " Duration: " << action.duration << '\n';
    // }
    cout << "Total actions: " << gen_actions.size() << '\n';

    Pose pose{100, 200, 10, 2};
    ObjectOrientedActionSpace oos (pose, 3, 10, 100, 3, 1, 5, 3);
    list<Object_Oriented_Action> oo_actions = oos.get_action_space();

    cout << "Total actions: " << oo_actions.size() << '\n';

}
